#1653 - Poprawki DOM w szablonie przedmiotu

jerzeuek opened the pull request:
Przyczyną występowania dodatkowego odstępu przed plusami w rzędzie "Liczba godzin" jest sama specyfika HTMLa - kiedy używamy `display: inline` (domyślna wartość dla m.in. `<span>`ów), to przeglądarka kondensuje wszystkie spacje (w tym newline'y występujące między elementami w kodzie) w jedną spację. 
Po rozpatrzeniu kilku możliwych rozwiązań najlepszym z nich jakie udało mi się wymyślić jest dodanie Djangowego tagu `{% spaceless %}`okalającego problematyczne elementy. Dzięki temu odstępy znikają, a kod nadal jest czytelny.
![obraz](https://github.com/iiuni/projektzapisy/assets/96109338/8c07a14d-3d8a-4a6f-9307-bda34142243d)

Dodatkowo zgodnie z sugestią we wszystkich rzędach gdzie były używane `<span>`y zostały one zastąpione listami wypunktowanymi z elementami. Został przy tym zachowany efekt wizualny.
![obraz](https://github.com/iiuni/projektzapisy/assets/96109338/5ab504c0-dadd-4832-aced-36675b11a745)

closes #1653

jerzeuek committed:
Usuwamy nadmiarowe białe znaki między plusami w rzędzie "Liczba godzin" używając Djangowego tagu spaceless.
Zamieniamy <span>y na elementy list wypunktowanych w rzędach "Liczba godzin", "Tagi" i "Grupy efektów kształcenia" zachowując przy tym oryginalny efekt wizualny.

jerzeuek committed:
Poprawiamy literówki, usuwamy niepotrzebne puste linie  w kodzie

jerzeuek committed:
Przywracamy usunięte puste linie pod koniec pliku

lgpawel commented on zapisy/apps/enrollment/courses/templates/courses/course_parts/course_head.html:
```diff
@@ -35,17 +42,25 @@ <h1 class="d-inline-block">
-                {% if course.hours_exercise_lab %}<span>{{ course.hours_exercise_lab }} (ćw-prac.)</span>{% endif %}
-                {% if course.hours_seminar %}<span>{{ course.hours_seminar }} (sem.)</span>{% endif %}
-                {% if course.hours_recap %}<span>{{ course.hours_recap }} (rep.)</span>{% endif %}
+                {% spaceless %}
+                <ul>
```
To pouczające, że jest taki znacznik jak `spaceless`, ale sięganie po niego wydaje się trochę desperackie, tj. wyobrażałbym sobie, że powinno się móc mieć jednocześnie "ładny", tj. niepozbawiony białych znaków (docelowy) HTML i oczywiście pożądany efekt wizualny. Czy nie da się tego osiągnąć przez odpowiednie atrybuty (m.in. `display`) na tym elemencie `<ul>` (i/lub zagnieżdżonych `<li>`)?

jerzeuek committed:
Przenosimy informacje o stylu CSS do jednego znacznika style.
Usuwamy tagi spaceless.
Zmieniamy właściwość CSS dla pola godzin z inline-block na flex.

jerzeuek commented:
Zgodnie z sugestią przeniosłem wszystkie informacje o stylu do jednego bloku ```<style>```. Udało się także w miarę łatwo uzyskać pożądany efekt usunięcia nadmiarowych białych znaków bez wcześniej użytego Djangowego znacznika ```{% spaceless %}``` - trzeba było zmienić właściwość CSS ```display``` dla pola z rozpiską godzin` z ```inline-block``` na ```flex```.

lgpawel commented on zapisy/apps/enrollment/courses/templates/courses/course_parts/course_head.html:
```diff
@@ -16,6 +16,25 @@ <h1 class="d-inline-block">
+
+        #hours ul {
+            display: flex;
+            padding: 0;
+            margin: 0;
+            list-style-type: none;
```
To wygląda na niepotrzebne w świetle takich samych definicji dla `#table-info ul`.

lgpawel commented:
No nie, nie tędy droga (nawet jeśli akurat dobrze wygląda), skoro nasze pozycjonowanie elementów nie ma w ogóle nic wspólnego z Flexboksem... Skoro nie `display-style` to może np. z `float` pokombinować?

Przy okazji proszę zrebase'ować gałąź tak, żeby można było oglądać zmiany naniesione na aktualn(iejsz)ą wersję systemu.

jerzeuek committed:
Usuwamy niepotrzebne własności, zamieniamy display na float

jerzeuek commented:
Szczerze mówiąc nie pomyślałem wcześniej z jakiegoś powodu o użyciu ```float```'a, ale faktycznie udało się z jego pomocą łatwo uzyskać zamierzony efekt.

Zgodnie z prośbą zrobiłem też rebase.

lgpawel commented on zapisy/apps/enrollment/courses/templates/courses/course_parts/course_head.html:
```diff
@@ -16,6 +16,22 @@ <h1 class="d-inline-block">
+            margin: 0;
+        }
+
+        #hours ul li {
```
Proszę jeszcze o eksperyment z taką zmianą (o ile nie ma jakiegoś ważnego powodu, żeby zostawić tak, jak jest):
```suggestion
        #table-info ul li {
```
– to nie ma znaczenia dla bieżącego efektu (ten już jest OK) ale byłoby to zgodne z intencją, jak widzimy **wszystkie** listy wypunktowane, jakie zdarzają się w tych tabelkach (choć inne już mają podobny wygląd dzięki dziedziczonemu po Bootstrapowej klasie `badge` własności `display: inline-block`).

jerzeuek commented:
Potestowałem proponowaną zmianę, jedyny większy efekt uboczny jaki zauważyłem to że badge'e są bliżej górnej krawędzi komórki i bliżej siebie.

Wariant z ``#hours ul li``:

![obraz](https://github.com/user-attachments/assets/657c991e-6bdc-4cf1-b041-025f1b0f19d7)

Wariant z ``#table-info ul li``:

![obraz](https://github.com/user-attachments/assets/f91adf97-f542-45f7-81d1-d63cd6b86a56)


Jeśli chcemy zachować tę modyfikację i jednocześnie zachować oryginalny wygląd pól z badge'ami, to możemy np. przypisać komórkom zawierającym badge'e pewną klasę (powiedzmy że np. ``badges``), a potem używając selektora ``:not()`` sprawić aby te komórki nie miały właściwości ``float: left``

Wtedy wyglądałoby to tak:
```css
#table-info ul li:not(.badges ul li) {
```

Oprócz tego drobiazgu innych problemów nie zauważono :)

jerzeuek committed:
Nadajemy klasę badges komórkom z badge'ami, naprawiamy wygląd komórek z <ul>ami

jerzeuek commented:
Znowu zrobiłem rebase, rozwiązałem również oczywiście powstały merge conflict
Po zmianach niektóre komórki miały wygląd niezgodny z oczekiwaniami (elementy w komórkach w których występuje ``<ul>``, czyli te z godzinami i badge'ami były przesunięte na prawo w porównaniu do reszty komórek), więc oczywiście zostało to przeze mnie naprawione.
